src/lib/insert-or-replace.js

```diff
@@ -1,12 +1,113 @@
-module.exports = (db, doc) =>
-  db.get(doc._id)
-    .then(existingDoc => doc._rev = existingDoc._rev)
-    .catch(e => {
-      if(e.status === 404) {
-        return;
+const mime = require('mime-types');
+const log = require('./log');
+
+const MAX_RETRY = 3;
+
+const getDoc = async (db, docId) => {
+  try {
+    return await db.get(docId);
+  } catch (e) {
+    if (e.status === 404) {
+      return null; // Document doesn't exist
+    }
+    throw e;
+  }
+};
+
+const putDoc = async (db, doc, existingRev = null) => {
+  const docToSave = { ...doc };
+  if (existingRev) {
+    docToSave._rev = existingRev;
+  } else {
+    delete docToSave._rev; // Ensure no _rev for new documents
+  }
+  return await db.put(docToSave);
+};
+
+const addDocAttachment = async (db, docId, attachmentName, attachment, currentRev, retries = MAX_RETRY) => {
+  if (retries < 0) {
+    throw new Error(`Failed to add attachment ${attachmentName} to ${docId} after retries`);
+  }
+
+  try {
+    const contentType = attachment.content_type || mime.lookup(attachmentName) || 'application/octet-stream';
+    const result = await db.putAttachment(docId, attachmentName, currentRev, attachment.data, contentType);
+    log.info(`Added attachment ${attachmentName} to ${docId}`);
+    return result;
+  } catch (err) {
+    if (err.status === 409) {
+      log.info(`Attachment conflict for ${attachmentName} in ${docId}, retrying`);
+      const latestDoc = await getDoc(db, docId);
+      if (!latestDoc) {
+        throw new Error(`Document ${docId} not found during attachment retry`);
       }
-      else {
-        throw e;
-      }
-    })
-    .then(() => db.put(doc));
+      return addDocAttachment(db, docId, attachmentName, attachment, latestDoc._rev, retries - 1);
+    }
+    throw err;
+  }
+};
+
+const handleAttachments = async (db, docId, attachments, initialRev) => {
+  let currentRev = initialRev;
+  for (const attachmentName of Object.keys(attachments)) {
+    const result = await addDocAttachment(db, docId, attachmentName, attachments[attachmentName], currentRev);
+    currentRev = result.rev; // Update rev for next attachment
+  }
+  return currentRev;
+};
+
+const handleLargeDocument = async (db, doc, retries = MAX_RETRY) => {
+  if (retries < 0) {
+    throw new Error(`Large document update failed for ${doc._id} after retries`);
+  }
+
+  try {
+    const attachments = doc._attachments;
+    const docWithoutAttachments = { ...doc };
+    delete docWithoutAttachments._attachments;
```

Comment:
To save roundabouts on generating attachments on the server, can you please try to preserve the non-media attachments here? These would be the form xml, model.xml and form.html. 
Can you please try this and check how it works server-side? We would ideally not see forms getting regenerated when a new media attachment is added.